packages/fxa-settings/src/lib/integrations/integration-factory.ts

```diff
@@ -0,0 +1,79 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import {
+  Integration,
+  OAuthRedirectIntegration,
+} from '../../models/integrations/base-integration';
+import { ModelDataStore, UrlQueryData } from '../model-data';
+import { ReachRouterWindow } from '../window';
+import { DefaultIntegrationFlags } from './integration-factory-flags';
+import { IntegrationFlags } from './interfaces/integration-flags';
+
+export class IntegrationFactory {
+  protected readonly data: ModelDataStore;
+  public readonly flags: IntegrationFlags;
+
+  constructor(opts: {
+    data?: ModelDataStore;
+    flags?: IntegrationFlags;
+    window: ReachRouterWindow;
+  }) {
+    const { window } = opts;
+    this.data = opts.data || new UrlQueryData(window);
+    this.flags =
+      opts.flags || new DefaultIntegrationFlags(new UrlQueryData(window));
+  }
+
+  /**
+   * Produces an integration object given the current data store's state.
+   * @returns An integration implementation.
+   */
+  getIntegration() {
+    const data = this.data;
+    const flags = this.flags;
+
+    let integration: Integration | undefined;
+
+    if (flags.isOAuth()) {
+      integration = this.createOAuthIntegration(data);
+    } else {
+      // TODO, port over `_getVerificationContext` and `_getContext` from app-start
+    }
+
+    // Run final validation. This will ensure that the all fields decorated with an @bind are in the
+    // the correct state.
+    // Commenting this out so that pages will stop erroring out when we don't have sufficient query params.
+    // This might be a TODO to restore this once we have all the data we need in the React app.
+    // integration?.validate();
```

Comment:
After seeing the code in use, I no longer think it's a good idea to do this automatically. The calling code can invoke a validate if it determines it's necessary. I think in most cases, we can just let validation run on accesses to the various fields.

Reply:
Thanks for letting me know this, I ended up refactoring this to just `return` which felt nicer.